### ISSUE TITLE
feature: Metadata

### ISSUE THREAD

**Issue by Otoris** [yask123/Instant-Music-Downloader#38]
What if this could grab the relevant file metadata + file title and apply it after it's been downloaded from YouTube automagically? Would definitely help keep an iTunes library clean.

**Comment by picklepicklepick** [comment]
Don't YT files have ID3 tags in them anyway?

**Comment by yask123** [comment]
It is possible to implement this.
I recently wrote which grabs just album art for all your downloaded songs, you can try it : http://iyask.me/fixalbumart/

**Comment by gamegoofs2** [comment]
I tried implementing youtube-dl's "--metadata-from-title" and "--add-metadata" but things weren't turning out well. I tried downloading Only Love by Mumford and Sons and it thought the title was "Only Love - Mumford and" and the artist was "Sons". 

I'd be willing to add this feature if anyone has any ideas of pulling the metadata from the youtube download.

**Comment by yask123** [comment]
Added this feature in commit  https://github.com/yask123/Instant-Music-Downloader/commit/71145bc3e0cf2eba5e7b2aa3082cd38d13ae66ad 
Update `sudo pip install instantmusic --upgrade` to try.
[logo]: http://i.imgur.com/ewtdSAR.png "Screenshot"

**Comment by yask123** [comment]
Not closing this issue since Its not 100% complete.
Currently it assumes the song tittle in youtube is written as `Artist name - Song name` which is true in most of the cases.
<img width="467" alt="ewtdsar" src="https://cloud.githubusercontent.com/assets/1812021/10454903/95224ed0-71d5-11e5-8ef1-fbe4442cdcf2.png">

Thus it currently doesn't write `album` name.

I plan to fix this in future.

**Comment by Otoris** [comment]
Something like gracenote or ACRCloud could be used to grab metadata. Only issue is it requires uploading at least a part of the song after it is downloaded by youtube-dl.

**Comment by ssundarraj** [comment]
Musicbrainz requires you to fingerprint and send the song. This could be used instead of uploading the entire song.

**Comment by yask123** [comment]
Yep , Musicbrainz cromaprint is awesome, integrating it in IMD :smile:

**Comment by dufferzafar** [comment]
AcoustID lookups only work for what's already in the database. And I don't think anyone has ever submitted IDs of youtube videos before, so that method probably won't work!

**Comment by yask123** [comment]
Damn :disappointed: 

Only alternative to do google search with song tittle and scrape meta data from websites ?

**Comment by ssundarraj** [comment]
@dufferzafar 
I think it might work for YouTube videos. As far as I know, it matches the fingerprint of the song. It doesn't really matter where you download it from as long as that song is in the database.

**Comment by yask123** [comment]
@ssundarraj  It didn't work, I tried multiple times with different songs downloaded from youtube.

**Comment by gamegoofs2** [comment]
I found something the other day. It was called EchoNest. It might work.

**Comment by yask123** [comment]
Tried this hack to grab `album name`

```python
album_name=''
        try:
            url = 'http://www.google.com/search?q='+qp(title)
            req = requests.get(url)
            response = req.content
            result=str(response)
            link_start=result.find('http://www.metrolyrics.com')
            link_end=result.find('html',link_start+1)
            link = result[link_start:link_end+4]
            lyrics_html = link
            a = requests.get(lyrics_html)
            html_doc=  a.content
            soup = BeautifulSoup(html_doc, 'html.parser')
            album_name = soup.find(id = "album-name-link").text
        except:
            print ('Album name not found')    
```
Tested multiple times with different songs, seems to work for most of them.
It would work for songs which have their lyrics in `metrolyrics.com`. From there I am scraping the meta data.
Eg: http://www.metrolyrics.com/21-guns-lyrics-green-day.html

```from 21st Century Breakdown```

**Comment by yask123** [comment]
Pushed it to the source: bd2d28bff904bd3d5e27ab4c9984f5cc5ea70c6a 

Have a look here: https://github.com/yask123/Instant-Music-Downloader/blob/master/instantmusic-0.1/bin/instantmusic#L123

Since I haven't updated the package yet, run
`$ subl `which instantmusic` ` 

Replace `subl` with whichever code editor you use, and replace all code with: https://github.com/yask123/Instant-Music-Downloader/blob/master/instantmusic-0.1/bin/instantmusic

**Comment by yask123** [comment]
Updated package.
Run `pip install instantmusic --upgrade` to try.

<img width="1108" alt="kniacdz" src="https://cloud.githubusercontent.com/assets/1812021/10542595/dcaad354-7436-11e5-8091-43adb96b8880.png">

**Comment by semaj** [comment]
@yask123 I notice in the above picture the song has the last character cut off.

I am experiencing this too - the actual filename is fine but whenever I open it in any media player (Plex, iTunes, cmus) it's cut off. Any idea what that could be?

**Comment by yask123** [comment]
@semaj  Thanks for reporting. Yes this issue exists. It's a minor bug (Indexes for slicing text are not accurate). 
I'll fix this bug after my exams get over (5th December).
